File: cpp-proio/src/reader.cc
#include <fcntl.h>
#include <sys/types.h>

#include <google/protobuf/io/gzip_stream.h>
#include <lz4.h>

#include "reader.h"
#include "writer.h"

using namespace proio;
using namespace google::protobuf;

Reader::Reader(int fd) {
    this->fd = fd;
    fileStream = new io::FileInputStream(fd);
    closeFDOnDelete = false;

    initBucket();
}

Reader::Reader(std::string filename) {
    fd = open(filename.c_str(), O_RDONLY);
    if (fd == -1) throw fileOpenError;
    fileStream = new io::FileInputStream(fd);
    closeFDOnDelete = true;

    initBucket();
}

Reader::~Reader() {
    pthread_mutex_lock(&mutex);
    pthread_mutex_unlock(&mutex);
    pthread_mutex_destroy(&mutex);

    if (bucketHeader) delete bucketHeader;
    delete compBucket;
    LZ4F_freeDecompressionContext(dctxPtr);
    delete bucket;
    delete fileStream;
    if (closeFDOnDelete) close(fd);
}

Event *Reader::Next() {
    pthread_mutex_lock(&mutex);
    Event *event = readFromBucket();
    pthread_mutex_unlock(&mutex);
    return event;
}

proto::BucketHeader *Reader::NextHeader() {
    pthread_mutex_lock(&mutex);
    readBucket(std::numeric_limits<uint64_t>::max());
    pthread_mutex_unlock(&mutex);
    return bucketHeader;
}

uint64_t Reader::Skip(uint64_t nEvents) {
    pthread_mutex_lock(&mutex);

    uint64_t bucketEventsLeft = 0;
    if (bucketHeader) {
        bucketEventsLeft = bucketHeader->nevents() - bucketEventsRead;
    }
    uint64_t nSkipped = 0;
    if (nEvents > bucketEventsLeft) {
        nSkipped += bucketEventsLeft;
        uint64_t n;
        while ((n = readBucket(nEvents - nSkipped)) > 0) nSkipped += n;
    }

    uint64_t nRead0 = bucketEventsRead;
    while (nSkipped < nEvents) {
        readFromBucket(false);
        if (bucketEventsRead == nRead0) break;
        nRead0 = bucketEventsRead;
        nSkipped++;
    }

    pthread_mutex_unlock(&mutex);
    return nSkipped;
}

void Reader::SeekToStart() {
    pthread_mutex_lock(&mutex);

    delete fileStream;
    if (lseek(fd, 0, SEEK_SET) == -1) throw seekError;
    fileStream = new io::FileInputStream(fd);

    readBucket();

    pthread_mutex_unlock(&mutex);
}

void Reader::initBucket() {
    compBucket = new BucketInputStream(0);
    bucketEventsRead = 0;
    bucketHeader = NULL;
    LZ4F_createDecompressionContext(&dctxPtr, LZ4F_VERSION);
    bucket = new BucketInputStream(0);

    pthread_mutex_init(&mutex, NULL);
}

Event *Reader::readFromBucket(bool doMerge) {
    if (bucket->BytesRemaining() == 0) readBucket();
    auto stream = new io::CodedInputStream(bucket);

    uint32_t protoSize;
    if (!stream->ReadLittleEndian32(&protoSize)) {
        delete stream;
        return NULL;
    }

    bucketEventsRead++;
    if (doMerge) {
        auto eventLimit = stream->PushLimit(protoSize);
        auto eventProto = new proto::Event;
        if (!eventProto->MergeFromCodedStream(stream) || !stream->ConsumedEntireMessage())
            throw deserializationError;
        stream->PopLimit(eventLimit);
        delete stream;
        auto event = new Event(eventProto);
        event->metadata = metadata;
        return event;
    } else {
        if (!stream->Skip(protoSize)) throw corruptBucketError;
        delete stream;
        return NULL;
    }
}

uint64_t Reader::readBucket(uint64_t maxSkipEvents) {
    auto stream = new io::CodedInputStream(fileStream);
    syncToMagic(stream);

    bucketEventsRead = 0;
    compBucket->Reset(0);
    bucket->Reset(0);

    uint32_t headerSize;
    if (!stream->ReadLittleEndian32(&headerSize)) {
        delete stream;
        return 0;
    }

    auto headerLimit = stream->PushLimit(headerSize);
    if (bucketHeader) delete bucketHeader;
    bucketHeader = new proto::BucketHeader;
    if (!bucketHeader->MergeFromCodedStream(stream) || !stream->ConsumedEntireMessage())
        throw deserializationError;
    stream->PopLimit(headerLimit);

    // Set metadata for future events
    for (auto keyValuePair : bucketHeader->metadata())
        metadata[keyValuePair.first] = std::make_shared<std::string>(keyValuePair.second);

    uint64_t bucketSize = bucketHeader->bucketsize();
    if (bucketHeader->nevents() > maxSkipEvents) {
        compBucket->Reset(bucketSize);
        if (!stream->ReadRaw(compBucket->Bytes(), bucketSize)) throw corruptBucketError;
        delete stream;
    } else {
        if (!stream->Skip(bucketSize)) throw corruptBucketError;
        delete stream;
        return bucketHeader->nevents();
    }

    switch (bucketHeader->compression()) {
        case LZ4: {
            bucket->Reset(dctxPtr, compBucket);
            break;
        }
        case GZIP: {
            io::GzipInputStream *gzipStream = new io::GzipInputStream(compBucket);
            bucket->Reset(*gzipStream);
            delete gzipStream;
            break;
        }
        default:
            BucketInputStream *tmpBucket = bucket;
            bucket = compBucket;
            compBucket = tmpBucket;
    }

    return 0;
}

uint64_t Reader::syncToMagic(io::CodedInputStream *stream) {
    uint8_t num;
    uint64_t nRead = 0;

    while (stream->ReadRaw(&num, 1)) {
        nRead++;

        if (num == magicBytes[0]) {
            bool goodSeq = true;

            for (int i = 1; i < 16; i++) {
                if (!stream->ReadRaw(&num, 1)) break;
                nRead++;

                if (num != magicBytes[i]) {
                    goodSeq = false;
                    break;
                }
            }
            if (goodSeq) break;
        }
    }
    return nRead;
}

BucketInputStream::BucketInputStream(uint64_t size) {
    offset = 0;
    bytes.resize(size);
    this->size = size;
}

BucketInputStream::~BucketInputStream() { ; }

inline bool BucketInputStream::Next(const void **data, int *size) {
    *data = &bytes[offset];
    *size = this->size - offset;
    offset = this->size;
    if (*size == 0) return false;
    return true;
}

inline void BucketInputStream::BackUp(int count) { offset -= count; }

inline bool BucketInputStream::Skip(int count) {
    offset += count;
    if (offset > size) {
        offset = size;
        return false;
    }
    return true;
}

inline int64 BucketInputStream::ByteCount() const { return offset; }

uint8_t *BucketInputStream::Bytes() { return &bytes[0]; }

uint64_t BucketInputStream::BytesRemaining() { return size - offset; }

void BucketInputStream::Reset(uint64_t size) {
    offset = 0;
    if (bytes.size() < size) bytes.resize(size);
    this->size = size;
}

uint64_t BucketInputStream::Reset(io::ZeroCopyInputStream &stream) {
    Reset(0);
    uint8_t *data;
    int size;
    while (stream.Next((const void **)&data, &size)) {
        offset = this->size;
        this->size += size;
        if (this->size > bytes.size()) bytes.resize(this->size);
        std::memcpy(&bytes[offset], data, size);
    }
    offset = 0;
    return this->size;
}

uint64_t BucketInputStream::Reset(LZ4F_dctx *dctxPtr, BucketInputStream *compBucket) {
    offset = 0;
    size = bytes.size();
    if (size == 0) Reset(minBucketWriteWindow);
    int srcSize;
    uint8_t *srcBuffer;
    compBucket->Next((const void **)&srcBuffer, &srcSize);
    int srcBytesRemaining = srcSize;
    int dstSize;
    uint8_t *dstBuffer;
    size_t hint;
    while (srcBytesRemaining > 0) {
        Next((const void **)&dstBuffer, &dstSize);
        size_t srcSizeTmp = srcSize;
        size_t dstSizeTmp = dstSize;
        hint = LZ4F_decompress(dctxPtr, dstBuffer, &dstSizeTmp, srcBuffer, &srcSizeTmp, NULL);
        if (LZ4F_isError(hint)) throw badLZ4FrameError;
        srcBytesRemaining -= srcSizeTmp;
        BackUp(dstSize - dstSizeTmp);
        if (offset == size) {
            size += minBucketWriteWindow;
            bytes.resize(size);
        }
        compBucket->BackUp(srcBytesRemaining);
        compBucket->Next((const void **)&srcBuffer, &srcSize);
    }
    size = offset;
    offset = 0;
    if (hint != 0) {
        LZ4F_resetDecompressionContext(dctxPtr);
        throw badLZ4FrameError;
    }
    return size;
}

File: java-proio/src/main/java/proio/Reader.java
package proio;

import java.io.ByteArrayInputStream;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.IOException;
import java.lang.Iterable;
import java.util.Iterator;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.GZIPInputStream;

import com.google.protobuf.ByteString;
import com.google.protobuf.CodedInputStream;
import com.google.protobuf.InvalidProtocolBufferException;
import net.jpountz.lz4.LZ4FrameInputStream;

public class Reader implements Iterable<Event>, Iterator<Event>
{
	private InputStream fileStream = null;
	private CodedInputStream stream = null;
    private CodedInputStream bucket = null;
    private Proto.BucketHeader bucketHeader = null;
    private int bucketEventsRead = 0;
    private int bucketIndex = 0;
    private Map<String, ByteString> metadata = new HashMap<String, ByteString>();

    private Event queuedEvent = null;

	public Reader(String filename)
			throws IOException {
		fileStream = new FileInputStream(filename);
		stream = CodedInputStream.newInstance(fileStream);
	}

	public Reader(InputStream fileStream)
			throws IOException {
		stream = CodedInputStream.newInstance(fileStream);
	}

    public Event next(boolean metaOnly)
            throws IOException {
        while (bucketHeader == null || bucketIndex >= bucketHeader.getNEvents()) {
            if (bucketHeader != null) {
                bucketIndex -= bucketHeader.getNEvents();
            }
            readHeader();
            if (bucketHeader == null) {
                return null;
            }
        }

        Event event = new Event();
        event.metadata = metadata;
        if (!metaOnly) {
            if (bucket == null) {
                readBucket();
            }
            try {
                readFromBucket(event);
            } catch (InvalidProtocolBufferException e) {
                readBucket();
                readFromBucket(event);
            }
        }

        return event;
    }

	public void close()
			throws IOException {
		if (fileStream != null) {
            fileStream.close();
        }
	}

	public Iterator<Event> iterator() {
		return this;
	}

	public boolean hasNext() {
        if (queuedEvent != null) {
            return true;
        }
        queuedEvent = next();
        if (queuedEvent != null) {
            return true;
        }
		return false;
	}

	public Event next() {
        if (queuedEvent != null) {
            Event returnEvent = queuedEvent;
            queuedEvent = null;
            return returnEvent;
        }
        try {
            return next(false);
        } catch (IOException e) {
            return null;
        }
	}

	public void remove() {}

    private void readHeader()
            throws IOException {
        bucket = null;
        bucketHeader = null;
        bucketEventsRead = 0;

        syncToMagic();

        int headerSize = stream.readRawLittleEndian32();

        int headerLimit = stream.pushLimit(headerSize);
        bucketHeader = Proto.BucketHeader.parseFrom(stream);
        stream.popLimit(headerLimit);

        for (Map.Entry<String, ByteString> entry : bucketHeader.getMetadataMap().entrySet()) {
            metadata.put(entry.getKey(), entry.getValue());
        }
    }

    private void readBucket()
            throws IOException {
        int bucketSize = (int) bucketHeader.getBucketSize();
        ByteArrayInputStream compBucket = new ByteArrayInputStream(stream.readRawBytes((int) bucketSize));

        switch (bucketHeader.getCompression()) {
            case GZIP:
                bucket = CodedInputStream.newInstance(new GZIPInputStream(compBucket));
                break;
            case LZ4:
                bucket = CodedInputStream.newInstance(new LZ4FrameInputStream(compBucket));
                break;
            default:
                bucket = CodedInputStream.newInstance(compBucket);
                break;
        }
    }

    private void readFromBucket(Event event)
            throws InvalidProtocolBufferException, IOException {
        while (bucketEventsRead <= bucketIndex) {
            int protoSize = bucket.readRawLittleEndian32();

            if (event != null && bucketEventsRead == bucketIndex) {
                int eventLimit = bucket.pushLimit(protoSize);
                event.eventProto = Proto.Event.parseFrom(bucket);
                bucket.popLimit(eventLimit);
            } else {
                bucket.skipRawBytes(protoSize);
            }

            bucketEventsRead++;
        }
        bucketIndex++;
    }

	private void syncToMagic()
			throws IOException {
		int nRead = 0;
        while (true) {
            byte thisByte = stream.readRawByte();

            if (thisByte == Writer.magicBytes[0]) {
                boolean goodSeq = true;

                for (int i = 1; i < 16; i++) {
                    thisByte = stream.readRawByte();

                    if (thisByte != Writer.magicBytes[i]) {
                        goodSeq = false;
                        break;
                    }
                }

                if (goodSeq) break;
            }
        }

		return;
	}
}
